File: src/main/java/com/intgrah/algorithms/sort/MergeSort.java
package com.intgrah.algorithms.sort;

import java.util.Comparator;

import static java.util.Arrays.copyOfRange;

public class MergeSort<T> extends Sorter<T> {

    public MergeSort(Comparator<T> ord) {
        super(ord);
    }

    @Override
    public void sort(T[] a) {
        for (int w = 1; w < a.length; w *= 2) {
            for (int r = a.length; r > 0; r -= 2 * w) {
                int m = r - w;
                if (m < 1) {
                    continue;
                }
                int l = m > w ? m - w : 0;
                T[] aux = copyOfRange(a, l, m);
                int lm = m - l;
                int i = 0;
                while (l < r) {
                    if (i < lm && (m == r || ord.compare(aux[i], a[m]) <= 0)) {
                        a[l++] = aux[i++];
                    } else {
                        a[l++] = a[m++];
                    }
                }
            }
        }
    }

}

File: sort/merge_sort.py
def merge_sort(a: list) -> None:
    aux = [0] * (len(a) // 2)
    w = 1
    while w < len(a):
        for r in range(len(a), 0, -2 * w):
            m = max(r - w, 0)
            l = max(m - w, 0)
            aux[: m - l] = a[l : m]
            lm = m - l
            i = 0
            while l < r:
                if i < lm and (m == r or aux[i] <= a[m]):
                    a[l] = aux[i]
                    i += 1
                else:
                    a[l] = a[m]
                    m += 1
                l += 1
        w *= 2
